modules/http/response-handler.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2019 Balabit
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "response-handler.h"
+
+HttpResult
+http_result_success(gpointer user_data)
+{
+  return HTTP_RESULT_SUCCESS;
+}
+
+HttpResult
+http_result_retry(gpointer user_data)
+{
+  return HTTP_RESULT_RETRY;
+}
+
+HttpResult
+http_result_drop(gpointer user_data)
+{
+  return HTTP_RESULT_DROP;
+}
+
+HttpResult
+http_result_disconnect(gpointer user_data)
+{
+  return HTTP_RESULT_DISCONNECT;
+}
+
+static void
+_free_entry(HttpResponseHandler *self)
+{
+  g_free(self);
+}
+
+HttpResponseHandlers *
+http_response_handlers_new(void)
+{
+  return g_hash_table_new_full(g_int_hash, g_int_equal, g_free, (GDestroyNotify)_free_entry);
+}
+
+void
+http_response_handlers_free(HttpResponseHandlers *self)
+{
+  g_hash_table_destroy(self);
+}
+
+static HttpResponseHandler *
+_clone(HttpResponseHandler *self)
+{
+  HttpResponseHandler *clone = g_new(HttpResponseHandler, 1);
+  *clone = *self;
+  return clone;
+}
+
+void
+http_response_handlers_insert(HttpResponseHandlers *self, HttpResponseHandler *response_handler)
+{
+  HttpResponseHandler *clone = _clone(response_handler);
+  gint *status_code = g_new(gint, 1);
```

Comment:
why do we need to allocate this gint? it is already embedded in the response_handler struct, so it could just be used from there and the key wouldn't need a free function.

Reply:
This one is a little bit tricky. When someone inserts a different value behind the same key, by default, hashtable does not replace the key. In the hashtable, the original key is maintained. As the value is overwritten, the old value is deallocated, and together with it, the embedded status code too. The next lookup will result in data loss or invalid read.

It is easy to run into, that's why I added a unit test for overwrite.

I added notes around the code for clarification and future readers.

Reply:
After thinking a little bit more about this, I might have an idea. There is a `g_hash_table_replace` that could be used, instead of `g_hash_table_insert`. According to documentation

> Inserts a new key and value into a GHashTable similar to g_hash_table_insert(). The difference is that if the key already exists in the GHashTable, it gets replaced by the new key.

And this is exactly the problem. If I use insert, the old key remains, that would be deallocated together with the value. Replace might work.